doc/plugin_server_nodeattestor_httppop.md

```diff
@@ -0,0 +1,53 @@
+# Server plugin: NodeAttestor "httppop"
+
+*Must be used in conjunction with the agent-side httppop plugin*
+
+The `httppop` plugin handshakes via http to ensure the agent is running on a valid
+dns name.
+
+The SPIFFE ID produced by the plugin is based on the dns name attested
+The SPIFFE ID has the form:
+
+```xml
+spiffe://<trust_domain>/spire/agent/httppop/<hostname>
+```
+
+| Configuration           | Description                                                                                                                                               | Default                             |
+|-------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------|
+| `dns_patterns`          | A list of regular expressions to apply to the hostname being attested. If none match, attestation will fail. If unset, all hostnames are allowed.         |                                     |
+| `allow_alternate_ports` | Set to true to allow ports other then 80 to be specified by the agent and honored during the handshake. If false, ports other then 80 will be rejected.   | false                               |
+| `allow_non_root_ports`  | Set to true to allow ports >= 1024 to be used by the agents with the advertised_port                                                                      | false                               |
```

Comment:
Is there danger in enabling these? Should we just always allow it?

Reply:
Some situations I can see (not exhaustive)

For allow_alternate_ports, if you have an internet facing spire-server, agents can make the server spend extra time/resources asking for callbacks to ports that are more likely to get blackholed I think. It could take longer for the server to decide to give up. Forcing it to be just one port controls the issue somewhat. Same with unintentially asking for an arbitrary port and have an intermediate firewall just block non port 80/443 and users wonder why the plugin is broken.

allow_non_root_ports being false adds an extra bit of security to things like NFS does. Say you have a shared unix box where multiple untrusted users can login, but only as their own users (no sudo root). They could http attest with a high port and get their own agent running on the node under their own user, when the system admin wants to use http attestation for the whole node with a root owned agent. These type nodes are common in HPC environments amongst others.

Reply:
> allow_non_root_ports being false adds an extra bit of security to things like NFS does. Say you have a shared unix box where multiple untrusted users can login, but only as their own users (no sudo root). They could http attest with a high port and get their own agent running on the node under their own user, when the system admin wants to use http attestation for the whole node with a root owned agent. These type nodes are common in HPC environments amongst others.

This is an important observation! It is the same problem as most CSP attestors , and there we work around it using TOFU. Should this attestor also have TOFU behavior for a given DNS name?

Reply:
TOFU sounds like a good option in addition to the allow_non_root_ports feature. It adds a different, and mostly complimentary feature I think.

For bare metal nodes, its more common to need to reuse the same hostname when reinstalling the node then in the cloud I think. Needing to clear out the name out of the spire server so it can be re-provisioned can painful and run into issues with automation. It may be worth it to some users, but not others to use the TOFU option due to this.

Also thinking forward, when spire has support for multiple attestors together so that say, a tpm plugin and an http_challenge plugin are both required, it would be desirable to not TOFU but reattest so that both the tpm and http challenge are valid with regular re-attestation, TOFU wouldn't work in that environment. Unless periodic reattestation could remove TOFU registration when reattesting I guess?